app/src/main/java/app/revenge/manager/installer/step/download/base/DownloadStep.kt

```diff
@@ -86,46 +117,51 @@ abstract class DownloadStep: Step() {
         }
 
         runner.logger.i("$fileName was not properly cached, downloading now")
-        var lastProgress: Float? = null
-        val result = downloadManager.download(url, destination) { newProgress ->
-            progress = newProgress
-            if (newProgress != lastProgress && newProgress != null) {
-                lastProgress = newProgress
-                runner.logger.d("$fileName download progress: ${(lastProgress!! * 100f).roundToInt()}%")
-            }
+
+        val isUsingMirror = url.toUri().host == null
+        var downloadUrl = if (isUsingMirror) {
+            preferenceManager.mirror.baseUrl + url
+        } else {
+            url
         }
 
-        when (result) {
-            is DownloadResult.Success -> {
-                try {
-                    runner.logger.i("Verifying downloaded file")
-                    verify()
-                    runner.logger.i("$fileName downloaded successfully")
-                } catch (t: Throwable) {
-                    mainThread {
-                        context.showToast(R.string.msg_download_verify_failed)
-                    }
-
-                    throw t
-                }
-                runner.logger.i("Moving $fileName to working directory")
-                destination.copyTo(workingCopy, true)
-            }
+        var successfulDownload = download(downloadUrl, destination, runner)
 
-            is DownloadResult.Error -> {
-                mainThread {
-                    context.showToast(R.string.msg_download_failed)
-                    runner.downloadErrored = true
+        // If the current mirror fails, try other mirrors
+        if (!successfulDownload && isUsingMirror) {
+            for (mirror in Mirror.entries) {
+                if (mirror == preferenceManager.mirror) continue
```

Comment:
Instead of this just do: (Mirror.entries - preferenceManager.mirror).forEach

Reply:
How can I break out of a forEach? The loop comes inside of the successfulDownload check

Reply:
Using return@